modules/core/src/test/java/org/apache/ignite/internal/processors/service/IgniteServiceDeploymentFailureTest.java

```diff
@@ -0,0 +1,216 @@
+
+    /** */
+    private static ClassLoader extClsLdr;
+
+    /** {@inheritDoc} */
+    @Override protected IgniteConfiguration getConfiguration(String igniteInstanceName) throws Exception {
+        IgniteConfiguration cfg = super.getConfiguration(igniteInstanceName);
+
+        cfg.setPeerClassLoadingEnabled(false);
+        cfg.setFailureHandler(new StopNodeFailureHandler());
+
+        return cfg;
+    }
+
+    /** {@inheritDoc} */
+    @Override protected void beforeTest() throws Exception {
+        super.beforeTest();
+
+        extClsLdr = getExternalClassLoader();
+    }
+
+    /** {@inheritDoc} */
+    @Override protected void afterTest() throws Exception {
+        stopAllGrids();
+
+        super.afterTest();
+
+        extClsLdr = null;
+    }
+
+    /**
+     *  Tests that deploying a service with a missing class causes a ServiceDeploymentException.
+     *
+     * @throws Exception If failed.
+     * */
+    @Test
+    public void testFailWhenClassNotFound() throws Exception {
+        IgniteEx srv = startGrid(getConfiguration("server"));
+        IgniteEx cli = startClientGrid(1);
+
+        ServiceConfiguration svcCfg = new ServiceConfiguration()
+                .setName("TestDeploymentService")
+                .setService(((Class<Service>)extClsLdr.loadClass(NOOP_SERVICE_CLS_NAME)).getDeclaredConstructor().newInstance())
+                .setNodeFilter(((Class<IgnitePredicate<ClusterNode>>)extClsLdr.loadClass(NODE_FILTER_CLS_NAME))
+                        .getConstructor(UUID.class)
+                        .newInstance(cli.configuration().getNodeId()))
+                .setTotalCount(1);
+
+        assertThrowsWithCause(() -> cli.services().deploy(svcCfg), ServiceDeploymentException.class);
+
+        assertTrue(cli.services().serviceDescriptors().isEmpty());
+    }
+
+    /**
+     * Service that throws an exception in init.
+     * */
+    private static class InitThrowingService implements Service {
+        /** */
+        public static final AtomicInteger initCounter = new AtomicInteger();
+
+        /** {@inheritDoc} */
+        @Override public void init() throws Exception {
+            initCounter.incrementAndGet();
+            throw new Exception("Service init exception");
+        }
+    }
+
+    /**
+     * NoopService for testing.
+     * */
+    private static class NoopService implements Service {
+        /** */
+        public static final AtomicInteger initCounter = new AtomicInteger();
+
+        /** {@inheritDoc} */
+        @Override public void init() throws Exception {
+            initCounter.incrementAndGet();
+        }
+    }
+
+    /**
+     * Tests that service descriptors are clear after an attempt of deploying
+     * a service which throws an exception during initialization
+     *
+     * @throws Exception If failed.
+     * */
+    @Test
+    public void testServerDescriptorsOfFailedServices() throws Exception {
+        final int NOOP_SRVC_TOTAL_CNT = 20;
```

Comment:
variables must be in camelCase -> noopSrvcTotalCnt here and below.

Reply:
fixed